src/notebooks/record_dedupe.py

```diff
@@ -0,0 +1,111 @@
+from typing import Callable
+import pandas as pd
+
+PATH_TO_DATA = '/Users/Brandon/Downloads/Archive/Patient-1.ndjson'
+
+def make_test_dedupe_data(df: pd.DataFrame):
+    '''
+    Turn a bundle of patient data into a small test set for verifying
+    efficacy of deduplication. There will be 6 unique records here,
+    4 of which are completely dissimilar, one which will have three
+    duplicated versions of varying changes, and one which will be a 
+    near match for the duplicates but is off by one character.
+    '''
+    
+    # Full match case: everything the same, should be no merge conflicts
+    df.loc[6, :] = df.loc[0, :]
+
+    # Most match case: one field blank, one present
+    df.loc[7, :] = df.loc[0, :]
+    df.loc[7, 'gender'] = ''
+
+    # Multi-match case: store both in ndjson
+    df.loc[8, :] = df.loc[0, :]
+    df.loc[8, 'address'] = df.loc[10, 'address']
+
+    # Near-match case: not actually a dupe, but off by one char
+    df.loc[9, :] = df.loc[0, :]
+    dob_list = df.loc[9, 'birthDate'].split('-')
+    dob_list[1] = str((int(dob_list[1]) + 1) % 12)
+    df.loc[9, 'birthDate'] = '-'.join(dob_list)
+
+    return df[:10]
+
+
+def create_linking_identifier(df: pd.DataFrame):
+    '''
+    Creates a basic identifier used to check duplicates in an incoming
+    batch of data. The basic identifier for the prototype is
+      FIRSTNAME-LASTNAME-YYYY-MM-DD
+    This identifier is joined to the dataframe proper for use in future
+    computation, as well as potential future use in generating/referencing
+    a master ID.
+    '''
+    df['last_name'] = df['name'].apply(lambda x: x[0]['family'] if type(x[0]['family']) is str else '-'.join(x[0]['family']))
```

Comment:
$0.02: would it be simpler to collate to a list? Something like `df['match-id'] = (df['name']['family'], df['name']['given'], ...)`

Reply:
We should find a FHIR compliant way to store the new identifier.

Reply:
Related to this to should the new identifier be added to the other resources belonging to this patient. This is not possible with data currently loaded, but probably worth considering before import to FHIR server.

Reply:
This has now been addressed with the updated PR. We no longer treat the concatenation of separate fields as the `master_id` (which is beyond the scope of this PR at present, though we will likely need to do a follow up to generate uuid at some point). Moreover, I've added a function that uses the found duplicates in a patient batch to alter the ID pointers in any provided resource to point to the de-duped "master" version of a coalesced record. It's inherently FHIR compliant because, in accordance with our discussion, we're just overwriting make believe numbers anyway. No risk of putting stuff in that the server won't recognize, while also letting us keep a quick and efficient linking system.